fe1-web/src/features/evoting/network/messages/__tests__/CastVote.test.ts

```diff
@@ -0,0 +1,238 @@
+
+  const mockVoteVotes1 = new Set([0]);
+  const mockVoteVotes2 = new Set([1, 0]);
+
+  const mockVoteId1 = CastVote.computeVoteId(election, 0, mockVoteVotes1);
+  const mockVoteId2 = CastVote.computeVoteId(election, 1, mockVoteVotes2);
+
+  mockVoteObject1 = {
+    id: mockVoteId1.toString(),
+    question: mockQuestionId1.valueOf(),
+    vote: [0],
+  };
+
+  mockVoteObject2 = {
+    id: mockVoteId2.toString(),
+    question: mockQuestionId2.valueOf(),
+    vote: [0],
+  };
+
+  const mockVotes = [mockVoteObject1];
+
+  sampleCastVote = {
+    object: ObjectType.ELECTION,
+    action: ActionType.CAST_VOTE,
+    lao: mockLaoIdHash,
+    election: electionId,
+    created_at: TIMESTAMP,
+    votes: mockVotes,
+  };
+
+  CastVoteJson = `{
+    "object": "${ObjectType.ELECTION}",
+    "action": "${ActionType.CAST_VOTE}",
+    "lao": "${mockLaoIdHash}",
+    "election": "${electionId}",
+    "created_at": ${TIMESTAMP},
+    "votes": ${JSON.stringify(mockVotes)}
+  }`;
+};
+
+beforeAll(() => {
+  configureTestFeatures();
+  initializeData();
+  OpenedLaoStore.store(mockLao);
+});
+
+describe('CastVote', () => {
+  it('should be created correctly from Json', () => {
+    expect(new CastVote(sampleCastVote as MessageDataProperties<CastVote>)).toBeJsonEqual(
+      sampleCastVote,
+    );
+
+    const temp = {
+      object: ObjectType.ELECTION,
+      action: ActionType.CAST_VOTE,
+      lao: mockLaoIdHash,
+      election: electionId,
+      created_at: TIMESTAMP,
+      votes: [mockVoteObject1, mockVoteObject2],
+    };
+    expect(new CastVote(temp)).toBeJsonEqual(temp);
+  });
+
+  it('should be parsed correctly from Json', () => {
+    const obj = JSON.parse(CastVoteJson);
+    expect(CastVote.fromJson(obj)).toBeJsonEqual(sampleCastVote);
+  });
+
+  it('fromJson should throw an error if the Json has incorrect action', () => {
+    const obj = {
+      object: ObjectType.ELECTION,
+      action: ActionType.NOTIFY_ADD,
+      lao: mockLaoIdHash.toString(),
+      election: electionId.toString(),
+      created_at: parseInt(TIMESTAMP.toString(), 10),
+      votes: [mockVoteObject1, mockVoteObject2],
+    };
+    const createFromJson = () => CastVote.fromJson(obj);
+    expect(createFromJson).toThrow(ProtocolError);
+  });
+
+  it('fromJson should throw an error if the Json has incorrect object', () => {
+    const obj = {
+      object: ObjectType.CHIRP,
+      action: ActionType.CAST_VOTE,
+      lao: mockLaoIdHash.toString(),
+      election: electionId.toString(),
+      created_at: parseInt(TIMESTAMP.toString(), 10),
+      votes: [mockVoteObject1, mockVoteObject2],
+    };
+    const createFromJson = () => CastVote.fromJson(obj);
+    expect(createFromJson).toThrow(ProtocolError);
+  });
+
+  describe('constructor', () => {
+    it('should throw an error if election is undefined', () => {
+      const createWrongObj = () =>
+        // @ts-ignore. Here was pass data to the constructor that is missing a field even though ts would require it
```

Comment:
I guess @Tyratox wants to try and break the message parsing.
I don't like it either, but I don't see how we can avoid the ts-ignore in this scenario 🤷

Reply:
We already did it in other tests and the IDE (and ESLint) didn't seem to complain, at least on my end.

Reply:
Yes but I changed the typing of the constructors to now allow undefined inputs anymore. The only place this can still happen is in the `fromJson` function where we work with an `any` type. Since I still wanted to test these cases a ts-ignore was necessary, otherwise VSCode reports the following error:

```
Argument of type '{ lao: Hash; created_at: Timestamp; votes: Vote[]; }' is not assignable to parameter of type 'MessageDataProperties<CastVote>'.
  Property 'election' is missing in type '{ lao: Hash; created_at: Timestamp; votes: Vote[]; }' but required in type 'MessageDataProperties<CastVote>'.ts(2345)
[CastVote.ts(17, 19): ]()'election' is declared here.
```